include/cifer/abe/dippe.h

```diff
@@ -0,0 +1,220 @@
+    ECP2_BN254* Ki;
+} cfe_dippe_user_secret_key;
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk);
+
+/**
+ * cfe_dippe represents the DIPPE scheme
+ */
+typedef struct cfe_dippe {
+    unsigned int assump_size;
+    mpz_t        p;
+    ECP_BN254*   g1_A;
+    ECP_BN254*   g1_UA;
+} cfe_dippe;
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe);
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_setup(cfe_dippe *dippe, unsigned int assump_size);
+
+/**
+ * Initializes an authority and generates its public and private key
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param pk A pointer to a cfe_dippe_pub_key struct; Represents the public key that is about to be populated
+ * @param sk A pointer to a cfe_dippe_sec_key struct; Represents the secret key that is about to be populated
+ */
+void cfe_dippe_authsetup(cfe_dippe *dippe, cfe_dippe_pub_key *pk, cfe_dippe_sec_key *sk);
+
+/**
+ * Encrypts a given message under the provided policy
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param cipher A pointer to a cfe_dippe_cipher struct; Represents the ciphertext that is about to be populated
+ * @param pks An array containing references to cfe_dippe_pub_key structs;
+ * @param pks_len Length of the public key array
+ * @param pol A reference to a cfe_vec struct; Represents the policy under which the message will be encrypted
+ * @param msg A reference to a FP12_BN254 struct; Represents the message that is about to be encrypted
+ * @return Error code
+ */
+cfe_error cfe_dippe_encrypt(cfe_dippe *dippe, cfe_dippe_cipher *cipher, cfe_dippe_pub_key *pks[], size_t pks_len, cfe_vec *pol, FP12_BN254 *msg);
+
+/**
+ * Used with an authority's secret key to create a new user secret key for a given attribute vector
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct; Represents the user secret key that is about to be populated
+ * @param usk_id Index of the given attribute vector for which a user secret key will be created
+ * @param pks An array containing references to cfe_dippe_pub_key structs
+ * @param pks_len Length of the public key array
+ * @param attrs A reference to a mpz_t struct; Represents the attribute vector
+ * @param gid String that represents a unique user; Required for collusion prevention
+ * @return Error code
+ */
+cfe_error cfe_dippe_keygen(cfe_dippe *dippe, cfe_dippe_user_secret_key *usk, unsigned int usk_id, cfe_dippe_pub_key *pks[], unsigned int pks_len, cfe_dippe_sec_key *sk, mpz_t *attrs, char gid[]);
```

Comment:
You don't need to pass `mpz_t` variables to functions as pointers - even in the GMP library, they always pass them as values.

Reply:
Good to know, I fixed it.